src/connectionpools.jl

```diff
@@ -0,0 +1,259 @@
+module ConnectionPools
+
+export Pod, Pool, acquire, release
+
+import Base: acquire, release
+
+connectionid(x) = objectid(x)
+
+"""
+    ConnectionTracker(conn::T)
+
+Wraps a `Connection` of type `T`.
+A `Connection` object must support the following interface:
+  * `isopen(x)`: check if a `Connection` object is still open and can be used
+  * `close(x)`: close a `Connection` object; `isopen(x)` should return false after calling `close`
+  * `ConnectionPools.connectionid(x)`: optional method to distinguish `Connection` objects from each other; by default, calls `objectid(x)`, which is valid for `mutable struct` objects
+
+The `idle` field is a timestamp to track when a `Connection` was returned to a `Pod` and became idle.
+
+The `count` field keeps track of how many times the connection has been used.
+"""
+mutable struct ConnectionTracker{T}
+    conn::T
+    idle::Float64
+    count::Int
+end
+
+ConnectionTracker(conn::T) where {T} = ConnectionTracker(conn, time(), 1)
+
+"""
+    Pod(T; max::Int, idle::Int, reuse::Int)
+
+A threadsafe object for managing a pool of and the reuse of `Connection` objects (see [`ConnectionTracker`](@ref)).
+
+A Pod manages a collection of `Connection`s and the following keyword arguments allow configuring the management thereof:
+
+  * `max::Int=typemax(Int)`: controls the max # of currently acquired `Connection`s allowed
+  * `idle::Int=typemax(Int)`: controls the max # of seconds a `Connection` may be idle before it should be closed and not reused
+  * `reuse::Int=typemax(Int)`: controls the max # of times a `Connection` may be reused before it should be closed
+
+After creating a `Pod`, `Connection`s can be acquired by calling [`acquire`](@ref) and MUST
+be subsequently released by calling [`release`](@ref).
+"""
+struct Pod{T}
+    # this lock/condition protects the `conns` Vector and `active` Dict
+    # no changes to either field should be made without holding this lock
+    lock::Threads.Condition
+    conns::Vector{ConnectionTracker{T}}
+    active::Dict{Any, ConnectionTracker{T}}
+    max::Int
+    idle::Int
+    reuse::Int
+end
+
+const MAX = typemax(Int)
+
+function Pod(T; max::Int=MAX, idle::Int=MAX, reuse::Int=MAX)
+    return Pod(Threads.Condition(), ConnectionTracker{T}[], Dict{Any, ConnectionTracker{T}}(), max, idle, reuse)
+end
+
+# check if an idle `Connection` is still valid to be reused
+function isvalid(pod::Pod{C}, conn::ConnectionTracker{C}) where {C}
```

Comment:
So after removing the "reuse" tracking, I feel like the function lives up more to the predicate `isvalid` name; it only closes in the case that the idle_timeout has been reached, but otherwise is mainly check if the connection is still open, so I'm inclined to keep the name.

Reply:
I'd say closing "sometimes" is still extremely surprising!  But ok :)